internal/bzlmod/go_mod.bzl

```diff
@@ -240,16 +326,55 @@ def sums_from_go_mod(module_ctx, go_mod_label):
     """
     _check_go_mod_name(go_mod_label.name)
 
-    # We go through a Label so that the module extension is restarted if go.sum
+    return parse_sumfile(module_ctx, go_mod_label, "go.sum")
+
+def sums_from_go_work(module_ctx, go_work_label):
+    """Loads the entries from a go.work.sum file given a go.work label.
+
+    Args:
+        module_ctx: a https://bazel.build/rules/lib/module_ctx object
+            passed from the MODULE.bazel call.
+        go_work_label: a Label for a `go.work` file. This label is used
+            to find the associated `go.work.sum` file.
+
+    Returns:
+        A Dict[(string, string) -> (string)] is returned where each entry
+        is defined by a Go Module's sum:
+            (path, version) -> (sum)
+    """
+    _check_go_work_name(go_work_label.name)
+
+    # next we need to test if the go.work.sum file exists, this is a little tricky so we use an indirect approach:
+
+    # 1. convert go_work_label into a path
+    go_work_path = module_ctx.path(go_work_label)
+
+    # 2. use the go_work_path to create a path for the heisen go.work.sum file
+    maybe_go_work_sum_path = go_work_path.get_child("../go.work.sum")
+
+    # 3. check for it's existence
+    if maybe_go_work_sum_path.exists:
+        return parse_sumfile(module_ctx, go_work_label, "go.work.sum")
+    else:
+        # 4. if go.work.sum does not exist, we should watch it incase it appears in the future
+        if hasattr(module_ctx, "watch_tree"):
+            # module_ctx.watch_tree is only available in bazel >= 7.1
+            module_ctx.watch_tree(maybe_go_work_sum_path)
```

Comment:
@fmeum here is my, seemingly working, attempt at implementing https://github.com/bazelbuild/bazel-gazelle/pull/1731#discussion_r1564615266

I've tested this manually a little, with and without go.work.sum files, and it appears to work. But it could use another set of eyes to sanity check.